xmtp/src/association.rs

```diff
@@ -0,0 +1,109 @@
+use crate::types::Address;
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+use xmtp_crypto::signature::{RecoverableSignature, SignatureError, SigningKey};
+use xmtp_crypto::utils::{self, eth_address};
+
+#[derive(Debug, Error)]
+pub enum AssociationError {
+    #[error("bad signature")]
+    BadSignature(#[from] SignatureError),
+    #[error("Association text mismatch")]
+    TextMismatch,
+    #[error(
+        "Address mismatch in Association: Provided:{provided_addr:?} != signed:{signing_addr:?}"
+    )]
+    AddressMismatch {
+        provided_addr: Address,
+        signing_addr: Address,
+    },
+    #[error("unknown association error")]
+    Unknown,
+}
+
+#[derive(Serialize, Deserialize, PartialEq, Debug)]
+pub enum AssociationText {
+    Static { addr: Address, key_bytes: Vec<u8> },
+}
+
+impl AssociationText {
+    pub fn get_address(&self) -> Address {
+        match self {
+            Self::Static { addr, .. } => addr.clone(),
+        }
+    }
+
+    pub fn text(&self) -> String {
+        match self {
+            Self::Static { addr, key_bytes } => gen_static_text_v1(addr, key_bytes),
+        }
+    }
+
+    pub fn is_valid(&self, addr: &str, key_bytes: &[u8]) -> Result<(), AssociationError> {
+        if self.text() == gen_static_text_v1(addr, key_bytes) {
+            return Ok(());
+        }
+
+        Err(AssociationError::TextMismatch)
+    }
+
+    pub fn new_static(addr: String, key_bytes: Vec<u8>) -> Self {
+        AssociationText::Static {
+            addr,
+            key_bytes,
+        }
+    }
+}
+
+fn gen_static_text_v1(addr: &str, key_bytes: &[u8]) -> String {
+    format!("XMTPv3 Link:{addr} -> keyBytes:{}", &hex::encode(key_bytes))
+}
+
+#[derive(Serialize, Deserialize, PartialEq, Debug)]
+pub struct Association {
+    pub text: AssociationText,
+    pub signature: RecoverableSignature,
+}
+
+impl Association {
+    pub fn new(text: AssociationText, signature: RecoverableSignature) -> Self {
+        Self { text, signature }
+    }
+
+    pub fn is_valid(&self, key_bytes: &[u8]) -> Result<(), AssociationError> {
```

Comment:
I like that idea. Might as well catch these things early so we don't have to check on every use.

Reply:
For tests, can't you just manually create the struct without calling the `::new` method?

Reply:
I made the members private so that no one creates one in production by accident. All Associations are valid by definitions. If Associations are guaranteed to be valid then theres no need to test for invalid ones 🤷